modules/apps/fragment/fragment-service/src/main/java/com/liferay/fragment/internal/deploy/auto/FragmentAutoDeployListener.java

```diff
@@ -290,27 +291,31 @@ else if (company != null) {
 	private static final Log _log = LogFactoryUtil.getLog(
 		FragmentAutoDeployListener.class);
 
-	@Reference
-	private CompanyLocalService _companyLocalService;
+	private static volatile FragmentsImporter _fragmentsImporter =
+		ServiceProxyFactory.newServiceTrackedInstance(
```

Comment:
@yuhai We should not use `ServiceProxyFactory` in a module.

Ideally, the listener for auto deploy should work in the same way as OSGi's bundle or service listeners. That is to say, even if the listener is initialized and registered later than an event, the listener can still be notified for that event, so you don't need to take care of the registration order.

Is it possible to make our auto deployer support such feature?

Reply:
cc @tinatian

Reply:
> @yuhai We should not use `ServiceProxyFactory` in a module.
> 
> Ideally, the listener for auto deploy should work in the same way as OSGi's bundle or service listeners. That is to say, even if the listener is initialized and registered later than an event, the listener can still be notified for that event, so you don't need to take care of the registration order.
> 
> Is it possible to make our auto deployer support such feature?

@dantewang, yes. I have a idea to create one module **portal-auto-deploy-listener-deployer**. In this module, we may create one component **AutoDeployListenerFileInstaller** as **FileInstaller.class** published. And then move the [logic ](https://github.com/yuhai/liferay-portal/blob/LPS-173157/portal-kernel/src/com/liferay/portal/kernel/deploy/auto/AutoDeployDir.java#L55-L64) from **AutoDeployDir** to **AutoDeployListenerFileInstaller**

This refer to LicenseInstaller.class and use DirectoryWatcher#run() logic. Is this good?

Reply:
@yuhai Yeah, just migrate `FragmentAutoDeployListener` to `FragmentFileInstaller`. We also have `BatchEngineAutoDeployListener` which should also be migrated.

After that, please remove the service tracker and related logic from `AutoDeployDir`. We this support wasn't added to support such usages.